### PR TITLE
fix: remove invalid feature gates

### PR DESCRIPTION
**Pull request by christinalau0**
**Reason for Change**:

Removes the invalid feature flag VolumeSnapshotDataSource in apiserver for k8s versions >= 1.22. Provides common util method for other invalid feature flags.

**Issue Fixed**:

This [original PR](https://github.com/kubernetes/kubernetes/pull/101531) resulted in the feature flag VolumeSnapshotDataSource becoming invalid. Customer with existing api-models containing this flag got broken deployments and upgrades. This PR will automatically remove the feature flag even if the customer specifies it, if k8s version >= 1.22. 

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [x] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: pkg/api/defaults-apiserver.go [hunk 1]
```diff
@@ -164,6 +164,13 @@ func (cs *ContainerService) setAPIServerConfig() {
 	if common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.20.0-alpha.1") && o.KubernetesConfig.APIServerConfig["--service-account-issuer"] == "kubernetes.default.svc" {
 		o.KubernetesConfig.APIServerConfig["--service-account-issuer"] = "https://kubernetes.default.svc.cluster.local"
 	}
+
+	outdatedFeatureGates := []string{}
+	// Remove --feature-gate VolumeSnapshotDataSource starting with 1.22
+	if common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.22.0-alpha.1") {
+		outdatedFeatureGates = append(outdatedFeatureGates, "VolumeSnapshotDataSource")
```
Comment (xdarsie) [inline pkg/api/defaults-apiserver.go:171]: could you please add the link to the PR?
https://github.com/kubernetes/kubernetes/pull/101531

File: pkg/api/defaults-apiserver.go [hunk 2]
```diff
@@ -164,6 +164,13 @@ func (cs *ContainerService) setAPIServerConfig() {
...
+	}
+	removeOutdatedFeatureGates(o.KubernetesConfig.APIServerConfig, outdatedFeatureGates)
```
Comment (xdarsie) [inline pkg/api/defaults-apiserver.go:173]: This random page indicates that not only api-server should be updated: 
https://d-nix.nl/2020/06/enable-feature-gates-on-a-live-kubernetes-cluster/

I am not sure the page is correct though. Please check which components needed the flag before 1.22 and clear those as well.

File: pkg/api/defaults-apiserver.go (hunk 1 above)
Comment (xdarsie) [reply pkg/api/defaults-apiserver.go:171]: here and as part of the PR description

File: pkg/api/defaults-apiserver.go [hunk 3]
```diff
@@ -164,6 +164,13 @@ func (cs *ContainerService) setAPIServerConfig() {
 	if common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.20.0-alpha.1") && o.KubernetesConfig.APIServerConfig["--service-account-issuer"] == "kubernetes.default.svc" {
 		o.KubernetesConfig.APIServerConfig["--service-account-issuer"] = "https://kubernetes.default.svc.cluster.local"
 	}
+
+	outdatedFeatureGates := []string{}
```
Comment (xdarsie) [inline pkg/api/defaults-apiserver.go:168]: not sure `outdated` is the term that reflects the issue, `invalid` may be slightly better?

File: pkg/api/defaults.go [hunk 4]
```diff
@@ -1129,6 +1129,23 @@ func combineValues(inputs ...string) string {
 	return mapToString(valueMap)
 }
 
+// remove specified outdated --feature-gates
```
Comment (xdarsie) [inline pkg/api/defaults.go:1132]: function comments should start with the function name
```
// removeOutdatedFeatureGates removes ...
```

File: pkg/api/defaults-apiserver_test.go [hunk 5]
```diff
@@ -493,6 +493,30 @@ func TestAPIServerFeatureGates(t *testing.T) {
 		t.Fatalf("got unexpected '--feature-gates' API server config value for k8s v%s: %s",
 			defaultTestClusterVer, a["--feature-gates"])
 	}
+
+	// test user-overrides, removal of VolumeSnapshotDataSource for k8s versions >= 1.22
+	cs = CreateMockContainerService("testcluster", defaultTestClusterVer, 3, 2, false)
+	cs.Properties.OrchestratorProfile.OrchestratorVersion = "1.22.0"
+	cs.Properties.OrchestratorProfile.KubernetesConfig.APIServerConfig = make(map[string]string)
+	a = cs.Properties.OrchestratorProfile.KubernetesConfig.APIServerConfig
+	a["--feature-gates"] = "VolumeSnapshotDataSource=true"
```
Comment (xdarsie) [inline pkg/api/defaults-apiserver_test.go:502]: please add a UT for "< v1.22" and ">= v1.22" without the feature gate

**Review by xdarsie — COMMENTED** [review]
Please add to the PR description why this change is necessary (what the feature gate breaks)

File: pkg/api/defaults-apiserver.go (hunk 2 above)
Comment (christinalau0) [reply pkg/api/defaults-apiserver.go]: Added to all relevant components

File: pkg/api/defaults-apiserver.go (hunk 3 above)
Comment (christinalau0) [reply pkg/api/defaults-apiserver.go]: Agreed, changed to invalid

File: pkg/api/defaults-apiserver.go [hunk 6]
```diff
@@ -164,6 +164,14 @@ func (cs *ContainerService) setAPIServerConfig() {
...
 		o.KubernetesConfig.APIServerConfig["--service-account-issuer"] = "https://kubernetes.default.svc.cluster.local"
 	}
+
+	invalidFeatureGates := []string{}
+	// Remove --feature-gate VolumeSnapshotDataSource starting with 1.22
+	// Reference: https://github.com/kubernetes/kubernetes/pull/101531
+	if common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.22.0-alpha.1") {
+		invalidFeatureGates = append(invalidFeatureGates, "VolumeSnapshotDataSource")
+	}
+	removeInvalidFeatureGates(o.KubernetesConfig.APIServerConfig, invalidFeatureGates)
```
Comment (CecileRobertMichon) [inline pkg/api/defaults-apiserver.go:174]: why are we doing this differently than what's already in place for other flags, ie. https://github.com/Azure/aks-engine/pull/4903/files#diff-f3e26129863d5da8cf5516c2d7abf5d9fe043bbb8119e076fd96b55ec2032414L151 ?

Comment (CecileRobertMichon) [reply pkg/api/defaults-apiserver.go:174]: I see, we're removing a specific feature gate within the flag, makes sense

File: pkg/api/defaults-scheduler.go [hunk 7]
```diff
@@ -36,4 +38,11 @@ func (cs *ContainerService) setSchedulerConfig() {
 	for key, val := range staticSchedulerConfig {
 		o.KubernetesConfig.SchedulerConfig[key] = val
 	}
+
+	invalidFeatureGates := []string{}
+	// Remove --feature-gate VolumeSnapshotDataSource starting with 1.22
+	if common.IsKubernetesVersionGe(o.OrchestratorVersion, "1.22.0-alpha.1") {
+		invalidFeatureGates = append(invalidFeatureGates, "VolumeSnapshotDataSource")
```
Comment (CecileRobertMichon) [inline pkg/api/defaults-scheduler.go:45]: is the `VolumeSnapshotDataSource` really present on all these components? My understanding was that it was only added to the apiserver

Comment (christinalau0) [reply pkg/api/defaults-scheduler.go:45]: Based on previous k8s version v1.19 documentation, `VolumeSnapshotDataSource` was a valid feature gate to specify in all these components.

**Review by CecileRobertMichon — APPROVED** [review]
lgtm

**Pull request merged**